### PR TITLE
Support for closed and unclosed paths

### PR DESCRIPTION
**Pull request by DivyanshiVashist**

### CODE REVIEW

File: src/toTransform.js [hunk 1]
```diff
@@ -62,17 +62,40 @@
     }
   }
 
+  function checkPathLoop (path) { 
+    var pathInput = path.input.replace(/[,\s]+$/g, '');
+
+    var lastPathInput = pathInput[pathInput.length - 1];
+    if(lastPathInput === 'z' || lastPathInput ==='Z') {
```
Comment (ewilligers) [inline src/toTransform.js]: === 'Z'
(add space)

Comment (ewilligers) [inline src/toTransform.js]: Or simply
return (lastPathInput === 'z' || lastPathInput === 'Z');

File: src/toTransform.js [hunk 2]
```diff
@@ -62,17 +62,40 @@
...
+    }
+
+    return false;
+  }
+
   function convertPathString (properties) {
     var offsetPath = internalScope.offsetPathParse(properties['offsetPath']);
-
-    var offsetDistance = internalScope.offsetDistanceParse(properties['offsetDistance']);
+    var closedLoop = checkPathLoop(offsetPath);
```
Comment (ewilligers) [inline src/toTransform.js]: perhaps isClosedLoop  to make clearer that the variable is true/false

File: src/toTransform.js [hunk 3]
```diff
@@ -62,17 +62,40 @@
...
+   var offsetDistance = internalScope.offsetDistanceParse(properties['offsetDistance']);
     if (offsetDistance === undefined) {
       offsetDistance = {value: 0, unit: 'px'};
     }
 
     pathElement.setAttribute('d', offsetPath.input);
 
     var offsetDistanceLength = getOffsetDistanceLength(offsetDistance, pathElement.getTotalLength());
+    if(offsetDistanceLength < 0) {
+      offsetDistanceLength = (offsetDistanceLength % pathElement.getTotalLength()) + pathElement.getTotalLength();
```
Comment (ewilligers) [inline src/toTransform.js]: This logic is only appropriate for closed loops. Can we avoided - see below.

File: src/toTransform.js [hunk 4]
```diff
@@ -62,17 +62,40 @@
...
+    }
+    
+    if(closedLoop) {
+      offsetDistanceLength = offsetDistanceLength % pathElement.getTotalLength();
+    }
```
Comment (ewilligers) [inline src/toTransform.js]: After updating using %, if resulting offsetDistanceLength < 0, add pathElement.getTotalLength()

File: src/toTransform.js [hunk 5]
```diff
@@ -62,17 +62,40 @@
...
+
+    if(!closedLoop && offsetDistanceLength > pathElement.getTotalLength()) {
```
Comment (ewilligers) [inline src/toTransform.js]: Perhaps have an else clause, which { checks if offsetDistanceLength < 0  else if  offsetDistanceLength > pathElement.getTotalLength() }

File: src/toTransform.js [hunk 6]
```diff
@@ -62,8 +62,20 @@
     }
   }
 
+  function checkPathLoop (path) {
```
Comment (alancutter) [inline src/toTransform.js]: isClosedLoop().

File: src/toTransform.js [hunk 7]
```diff
@@ -62,8 +62,20 @@
...
+    var pathInput = path.input.replace(/[,\s]+$/g, '');
+
+    var lastPathInput = pathInput[pathInput.length - 1];
+    if (lastPathInput === 'z' || lastPathInput === 'Z') {
+      return true;
+    }
+
+    return false;
```
Comment (alancutter) [inline src/toTransform.js]: Just return the if condition. There should be a toLower() or similarly named method as well.

File: src/toTransform.js [hunk 8]
```diff
@@ -73,8 +85,20 @@
...
+    if (offsetDistanceLength < 0) {
+      offsetDistanceLength = (offsetDistanceLength % pathElement.getTotalLength()) + pathElement.getTotalLength();
+    }
+
+    if (closedLoop) {
+      offsetDistanceLength = offsetDistanceLength % pathElement.getTotalLength();
+    }
+
+    if (!closedLoop && offsetDistanceLength > pathElement.getTotalLength()) {
+      offsetDistanceLength = pathElement.getTotalLength();
```
Comment (alancutter) [inline src/toTransform.js]: Store the total length in a variable since you use it so often.

File: src/toTransform.js [hunk 9]
```diff
@@ -73,8 +85,20 @@
...
     var offsetDistanceLength = getOffsetDistanceLength(offsetDistance, pathElement.getTotalLength());
+    if (offsetDistanceLength < 0) {
+      offsetDistanceLength = (offsetDistanceLength % pathElement.getTotalLength()) + pathElement.getTotalLength();
+    }
+
+    if (closedLoop) {
+      offsetDistanceLength = offsetDistanceLength % pathElement.getTotalLength();
+    }
+
+    if (!closedLoop && offsetDistanceLength > pathElement.getTotalLength()) {
```
Comment (alancutter) [inline src/toTransform.js]: This can be an else if.

File: src/toTransform.js [hunk 10]
```diff
@@ -73,8 +85,20 @@
     pathElement.setAttribute('d', offsetPath.input);
 
     var offsetDistanceLength = getOffsetDistanceLength(offsetDistance, pathElement.getTotalLength());
+    if (offsetDistanceLength < 0) {
+      offsetDistanceLength = (offsetDistanceLength % pathElement.getTotalLength()) + pathElement.getTotalLength();
+    }
+
+    if (closedLoop) {
+      offsetDistanceLength = offsetDistanceLength % pathElement.getTotalLength();
+    }
```
Comment (alancutter) [inline src/toTransform.js]: No need to do this adjustment if it's within the total length, you can treat it the same as !closedLoop.

File: test/offsetPathDistanceTest.js [hunk 11]
```diff
@@ -6,6 +6,104 @@
...
+        [
+                                    {at: 0, is: 'translate3d(0px, 0px, 0px) rotate(0deg)'},
+                                    {at: 1 / 6, is: 'translate3d(100px, 0px, 0px) rotate(0deg)'},
+                                    {at: 2 / 6, is: 'translate3d(100px, 100px, 0px) rotate(0deg)'},
+                                    {at: 3 / 6, is: 'translate3d(0px, 100px, 0px) rotate(0deg)'},
+                                    {at: 4 / 6, is: 'translate3d(0px, 100px, 0px) rotate(0deg)'},
+                                    {at: 5 / 6, is: 'translate3d(0px, 100px, 0px) rotate(0deg)'},
+                                    {at: 1, is: 'translate3d(0px, 100px, 0px) rotate(0deg)'}
+        ]
+      );
```
Comment (alancutter) [inline test/offsetPathDistanceTest.js]: These should have TODOs to fix the rotation.

File: src/toTransform.js [hunk 12]
```diff
@@ -74,7 +82,18 @@
...
     var offsetDistanceLength = getOffsetDistanceLength(offsetDistance, pathElement.getTotalLength());
 
+    if (closedLoop) {
+      if (offsetDistanceLength < 0) {
+        offsetDistanceLength = (offsetDistanceLength % pathElement.getTotalLength()) + pathElement.getTotalLength();
+      } else {
+        offsetDistanceLength = offsetDistanceLength % pathElement.getTotalLength();
+      }
+    } else if (offsetDistanceLength > pathElement.getTotalLength()) {
+      offsetDistanceLength = pathElement.getTotalLength();
```
Comment (alancutter) [inline src/toTransform.js]: You missed the total length variable nit.

**Pull request merged**